pkg/sql/alter_column_type_test.go

```diff
@@ -0,0 +1,521 @@
+	}
+}
+
+func TestVisibilityDuringAlterColumnType(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	defer setTestJobsAdoptInterval()()
+
+	ctx := context.Background()
+	swapNotification := make(chan struct{})
+	waitBeforeContinuing := make(chan struct{})
+	params, _ := tests.CreateTestServerParams()
+	params.Knobs = base.TestingKnobs{
+		SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
+			RunBeforeComputedColumnSwap: func() {
+				// Notify the tester that the alter column type is about to happen.
+				swapNotification <- struct{}{}
+				// Wait for the tester to finish before continuing the swap.
+				<-waitBeforeContinuing
+			},
+		},
+	}
+	s, sqlDB, _ := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(ctx)
+
+	if _, err := sqlDB.Exec(
+		`SET enable_alter_column_type_general = true;`); err != nil {
+		t.Error(err)
+	}
+
+	if _, err := sqlDB.Exec(`
+CREATE DATABASE t;
+CREATE TABLE t.test (x INT);
+INSERT INTO t.test VALUES (1), (2), (3);
+`); err != nil {
+		t.Error(err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		if _, err := sqlDB.Exec(`ALTER TABLE t.test ALTER COLUMN x TYPE STRING`); err != nil {
+			t.Error(err)
+		}
+		wg.Done()
+	}()
+
+	<-swapNotification
+
+	row := sqlDB.QueryRow("SHOW CREATE TABLE t.test")
+	var scanName, create string
+	if err := row.Scan(&scanName, &create); err != nil {
+		t.Error(err)
+	}
+	expected := `CREATE TABLE test (
+	x INT8 NULL,
+	FAMILY "primary" (x, rowid)
+)`
+	if create != expected {
+		t.Fatalf("expected %s, found %s", expected, create)
+	}
+
+	// Let the schema change process continue.
+	waitBeforeContinuing <- struct{}{}
+	wg.Wait()
+
+	row = sqlDB.QueryRow("SHOW CREATE TABLE t.test")
+	if err := row.Scan(&scanName, &create); err != nil {
+		t.Error(err)
+	}
+	expected = `CREATE TABLE test (
+	x STRING NULL,
+	FAMILY "primary" (x, rowid)
+)`
+	if create != expected {
+		t.Fatalf("expected %s, found %s", expected, create)
+	}
+}
+
+// TestAlterColumnTypeFailureRollback tests running a primary key
+// change on a table created in the same transaction.
+func TestAlterColumnTypeFailureRollback(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	ctx := context.Background()
+	params, _ := tests.CreateTestServerParams()
+	s, sqlDB, kvDB := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(ctx)
+
+	if _, err := sqlDB.Exec(
+		`SET enable_alter_column_type_general = true;`); err != nil {
+		t.Error(err)
+	}
+
+	_, err := sqlDB.Exec(`
+CREATE DATABASE t;
+CREATE TABLE t.test (x STRING);
+INSERT INTO t.test VALUES ('1'), ('2'), ('HELLO');
+ALTER TABLE 
+t.test ALTER COLUMN x TYPE INT;
+`)
+	actual := err.Error()
+	expected := "pq: could not parse \"HELLO\" as type int: strconv.ParseInt: parsing \"HELLO\": invalid syntax"
+
+	if actual != expected {
+		t.Fatalf("expected %s, got %s", expected, actual)
+	}
+
+	desc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
+	// Expect to find two drop mutations, one to drop the Column, one to drop the
```

Comment:
I think a better test would be to continue polling the descriptor in a succeeds soon and ensure that we don't have any pending descriptors (i.e. the mutations get cleaned up).